ENH, DOC: Add support for interactive examples for NumPy with `jupyterlite-sphinx`

agriyakhetarpal opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

## Description

This PR adds support for WASM-powered interactive examples with [JupyterLite](https://jupyterlite.readthedocs.io/en/stable/) and [`jupyterlite-sphinx`](https://jupyterlite-sphinx.readthedocs.io/) in the NumPy documentation.

A similar enhancement was made for the documentation for SciPy (https://github.com/scipy/scipy/pull/20019) and that for PyWavelets (https://github.com/PyWavelets/pywt/pull/728).

### How to test this

> [!TIP]
> I have a GitHub Pages website deployed at https://agriyakhetarpal.github.io/numpy/ through which these changes can be experimented with. The CircleCI deployment does not render the examples properly, however, a fix is available as reported in https://github.com/numpy/numpy/pull/26745#issuecomment-2183812333 and will be fixed upstream in https://github.com/jupyterlite/jupyterlite-sphinx/pull/182.
>
> **Update, 24/07/2024:** this fix is now available in `jupyterlite-sphinx` version 0.16.2, and the CircleCI deployment should render the examples properly as well.

### Key changes (so far)

1. Configuration for `jupyterlite-sphinx` in `doc/source/conf.py` and `doc/Makefile`, coupled with changes to styling in `doc/source/_static/numpy.css`
2. Updates to the API documentation under files in `numpy/` under the "Examples" section in the docstrings for various modules

> [!NOTE]
> I have currently left out the following: `_datasource.py`, `_utils_impl.py`, `_version.py`, `_array_api_info.py`, `ctypeslib.py`, files under `numpy/_typing/`, and `numpy/_core/numerictypes.py`.

> [!TIP]
> The "NumPy user guide" section has not been modified at this time due to the size of the PR (in terms of the number of files modified). However, it can be modified or adapted to a Jupytext notebook similar to scipy/scipy#20303 to make it work well in an interactive set-up.

> [!IMPORTANT]
> The NumPy documentation will have to be rebuilt and reuploaded in https://github.com/numpy/doc/ for these changes to show up in the official documentation.

## What issue does this PR address?

Closes #25969

## Additional context

- Updates to NumPy in-tree are coupled with new releases for Pyodide, and Pyodide 0.27 as of 01/01/2025 brings NumPy version 2.0.2, which is slightly behind NumPy's latest release at the time of writing. Therefore, some examples or new features may not work as expected and can fail in some cases, for which a global warning text has been provided and displayed before each interactive example, and examples that are known to fail have been disabled via [the `try_examples.json` configuration file](https://jupyterlite-sphinx.readthedocs.io/en/stable/directives/try_examples.html#try-examples-json-configuration-file). A custom Pyodide distribution or a custom wheel of NumPy is not provided with this PR, and a solution is being explored with nightly WASM wheels for NumPy (established in gh-26134) via https://github.com/pyodide/pyodide/issues/4898.
- SciPy documentation: https://docs.scipy.org/doc/scipy/reference/index.html
- PyWavelets documentation: https://pywavelets.readthedocs.io/en/latest/ref/index.html
- xref: https://github.com/Quansight-Labs/czi-scientific-python-mgmt/issues/19

agriyakhetarpal committed:
DOC: Remove JupyterLite artifacts with make clean

agriyakhetarpal committed:
DOC: Git-ignore JupyterLite DB file

agriyakhetarpal committed:
DEP: Add interactive documentation requirements

agriyakhetarpal committed:
DOC: Add JupyterLite configuration

agriyakhetarpal committed:
DOC: Configure CSS for `TryExamples` buttons

Co-Authored-By: Albert Steppi <1953382+steppi@users.noreply.github.com>

agriyakhetarpal committed:
DOC: Configure a minimum height for the buttons

agriyakhetarpal committed:
REL: Add notes about interactive examples

agriyakhetarpal committed:
DOC, TST: Add `>>> import numpy as np` stub to docstrings

agriyakhetarpal committed:
DOC: Add `.. try_examples::` directive to API reference

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/_core/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/lib/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/ma/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/linalg/_linalg.py`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/fft/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/random/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/polynomial/`

agriyakhetarpal committed:
DOC, API: Add `>>> import numpy as np` stub to `numpy/exceptions.py`

agriyakhetarpal committed:
DOC: Remove generated notebooks with `make clean`

[skip cirrus] [skip azp]

charris commented:
This could use a release note.

agriyakhetarpal commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -458,16 +467,22 @@ array actually get written to disk.
->>> a = np.memmap('newfile.dat', dtype=float)
->>> print(a[10], a[30])
-10.0 30.0
+.. try_examples::
+
+  >>> import numpy as np
+
+  >>> a = np.memmap('newfile.dat', dtype=float, mode='w+', shape=1000)
+  >>> a[10] = 10.0
+  >>> a[30] = 30.0
+  >>> del a
+
+  >>> b = np.fromfile('newfile.dat', dtype=float)
+  >>> print(b[10], b[30])
+  10.0 30.0
+
+  >>> a = np.memmap('newfile.dat', dtype=float)
+  >>> print(a[10], a[30])
+  10.0 30.0
```
I haven't tested whether every example works because there are way too many, but I did test this and this is one of the examples that doesn't, and fails with the error

```
ValueError: mmap length is greater than file size
```

and reducing the shape from `1000` to `100` doesn't help: `OSError: [Errno 43] No such device`

agriyakhetarpal replied:
I'm not sure if there is a way to enable the Pyodide file system or interact with Pyodide directly through a Pythonic interface like this.

agriyakhetarpal commented on doc/source/_static/numpy.css:
Question on license compliance: since this file has been adapted from that for SciPy, do I need to add a link for attribution or make a mention of this in `LICENSES_bundled.txt`?

agriyakhetarpal committed:
REL: Add release note for gh-26745

[skip cirrus] [skip azp]

agriyakhetarpal commented:
> This could use a release note.

@charris, I addressed this in 484fe56bb37dc4eac5925f845294b0e49afc2ef2. Please let me know if a "Highlight" note would be okay or if I should switch this to `new_feature`.

agriyakhetarpal committed:
STY: Fix failing doctests

[skip cirrus] [skip azp]

agriyakhetarpal commented:
I forgot to skip Azure Pipelines and Cirrus in my previous commit – please don't mind the force-push, my apologies

agriyakhetarpal commented:
I notice that for some reason, the CircleCI build isn't able to load the examples, most likely because the files in `source/contents/` weren't copied or that `jupyterlite-pyodide-kernel` was missing. I shall investigate.

melissawm commented on doc/source/_static/numpy.css:
```diff
@@ -71,4 +72,33 @@ div.admonition-legacy>.admonition-title::after {
+
+.try_examples_button {
+  color: white;
+  background-color: #0054a6;
```
we should ideally be using colors from the PyData Sphinx theme color scheme, so we get any accessibility/dark mode/usability improvements as they come.

agriyakhetarpal replied:
Thanks – I'll try to experiment with the colours based on their scheme and update this

rgommers replied:
Can we just delete this whole file? I can't remember if there were any blockers for doing so. The pydata-sphinx-theme defaults should be better than what we have here.

agriyakhetarpal replied:
Using the PyData Theme defaults should be good, I will take a look at it (in another PR or here? Fine with either). However, we'll have to keep the changes for these buttons in specific, because they aren't styled particularly well by default:

<img width="669" alt="image" src="https://github.com/numpy/numpy/assets/74401230/92f94a9d-c42b-48b8-844f-1cc8aaaab365">

agriyakhetarpal replied:
In 0c7908f0485b72c2a1093286dc2cc1c92e64cb82, I decided to use `--pst-color-info` here, it seems to be the best suited for the examples and is easy on the eyes, besides `--pst-color-primary`.

agriyakhetarpal replied:
Resolving this conversation for now, please let me know if there's something I can help out with conforming to other accessibility guidelines.

melissawm reviewed (commented):
I also don't see the try examples button locally - should I do anything differently other than `spin docs`?

agriyakhetarpal commented:
> I also don't see the try examples button locally - should I do anything differently other than spin docs?

@melissawm, I'm not sure – I think you'll need to refresh your `conda`/`mamba` environment, to load the new dependencies or install `requirements/doc_requirements.txt`, perhaps? I am able to see the buttons locally.

melissawm commented:
Nevermind - I do see it now!

agriyakhetarpal committed:
Merge remote-tracking branch 'upstream/main' into interactive-docs

agriyakhetarpal committed:
DOC: inherit styles from PST, fix alignments

[skip cirrus] [skip azp]

melissawm reviewed (approved):
Overall, this looks great to me, so approving.

This is how it looks now: 
![Captura de imagem_20240621_165421](https://github.com/numpy/numpy/assets/3949932/1035c915-77dd-4859-8663-0b4237a26f9a)

We can argue about the looks of the button separately, but in the @agriyakhetarpal 's gh-pages deployment indeed everything is working. 

A couple of comments:
- Do we want to have a functional jupyterlite kernel in CircleCI to verify that this is working on every PR? I'm not sure that's necessary. So I wouldn't spend too much time on trying to get that working. Happy to hear other thoughts though.
- I know we haven't necessarily figured out the details yet, but I feel like we should have some mention of the examples and jupyterlite in general in the contribution and maintenance docs. I would either link or get inspiration from https://scipy.github.io/devdocs/dev/core-dev/index.html#wrapping-up (for the relese guide) and the "Interactive documentation" section on https://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#rendering-documentation-locally-with-sphinx . This can be done in a follow-up but in that case we should open an issue so we don't forget.

lesteve commented:
> Do we want to have a functional jupyterlite kernel in CircleCI to verify that this is working on every PR? I'm not sure that's necessary. So I wouldn't spend too much time on trying to get that working. Happy to hear other thoughts though.

This is interesting because a while ago there was some kind of limitation in CircleCI that made it not possible to have JupyterLite working at all. Looking at bit more it almost work now, except that you need to add `index.html` explicitly, don't ask me why but I would guess that this is some CircleCI quirk. I have opened a sphinx-gallery PR to fix this https://github.com/sphinx-gallery/sphinx-gallery/pull/1336. I am guessing you would need to do something similar in your sphinx extension?

For example in your case if you look at this [example](https://output.circle-artifacts.com/output/job/ed942196-1f96-46f5-aa5c-79b22ab1a838/artifacts/0/doc/build/html/reference/generated/numpy.fft.fft.html#numpy.fft.fft) and click on the button it will show 
![image](https://github.com/numpy/numpy/assets/1680079/69313d96-eb05-4a45-88b8-89e8b0d6a848)

Looking at the console it is trying to load `https://output.circle-artifacts.com/output/job/ed942196-1f96-46f5-aa5c-79b22ab1a838/artifacts/0/doc/build/html/lite/notebooks/?path=c442ac1e_f24f_43a1_89ea_3f59c0371488.ipynb`

Add a `index.html` to it and now it works:
`https://output.circle-artifacts.com/output/job/ed942196-1f96-46f5-aa5c-79b22ab1a838/artifacts/0/doc/build/html/lite/notebooks/index.html?path=c442ac1e_f24f_43a1_89ea_3f59c0371488.ipynb`

rgommers reviewed (commented):
About 90% of the diff here is adding `import numpy as np`, to make the documentation examples more self-contained and copy-paste-able. It seems fine to me, and I'd prefer to just do this now that there's one more good reason for it, rather than bikeshed it again to the extent done for SciPy in https://github.com/scipy/scipy/issues/13049.

However, this is technically kinda separate from supporting interactive docs, so I do have a preference for separating the two changes - because it makes it much easier to review and understand what we actually need to add for interactive docs.

The same goes for other example code cleanups; those are general docstring hygiene, which can be merged independently of adding `jupyterlite-sphinx` support. 

I know it's a bit of extra work, but I think it's useful to do - it will make much more clear what it takes to start using `jupyterlite-sphinx`. Perhaps it's not too hard to split this in two, because the interactive-specific changes are mostly in doc config files and `.rst` files, while all the general docstring cleanups are in `.py` files; so something like this should get you most of the way there:
```
git merge main
git reset main --mixed
git add numpy/*.py
git add numpy/*/*.py
git commit
```
Do you think you could try that @agriyakhetarpal?

Carreau commented on doc/release/upcoming_changes/26745.highlight.rst:
```diff
@@ -0,0 +1,8 @@
+-----------------------------------------------
+
+The NumPy documentation includes a number of examples that
+can now be run interactively using WebAssembly and Pyodide.
```
I know this is a stupid remark, but not everyone might know that `WebAssembly and Pyodide` means in your borwser.

```suggestion
can now be run interactively in your browser using WebAssembly and Pyodide.
Comment

```

Carreau commented on doc/source/_static/numpy.css:
```diff
@@ -1,10 +1,11 @@
 .navbar-brand {
-   height: 75px;
+  height: 75px;
 }
```
For another time, do you believe it would be worth having a pre-comit to format the css in the numpy repo ?

agriyakhetarpal replied:
I think so! I couldn't find many within the `pre-commit` framework that seem to be regularly maintained, but Node.js-based tools exist, such as https://github.com/stylelint/stylelint that one can use with `npm`, or better `npx`.

Carreau commented on doc/source/_static/numpy.css:
```diff
@@ -71,4 +72,32 @@ div.admonition-legacy>.admonition-title::after {
+  justify-content: flex-start;
+  gap: 10px;
+  margin-bottom: 20px;
+}
```
I think the `try_examples_outer_iframe`, need a 0.4em top margin to avoid a different in alignment and jumping content:
<img width="521" alt="Screenshot 2024-06-27 at 11 27 05" src="https://github.com/numpy/numpy/assets/335567/3e3840ca-2d55-4bbe-a108-5ada2d739d8c">

agriyakhetarpal replied:
Addressed in 0547d8f87cdea3e8003d8e99bf171388bd0412e5.

Carreau commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -648,12 +669,15 @@ N-dimensional enumeration
+
+  >>> import numpy as np
+  >>> for i, val in np.ndenumerate(a):
+  ...     if sum(i)%5 == 0: print(i, val)
```
While we are at it, is it worth putting the print on it's own line ?

agriyakhetarpal replied:
Addressed in 7b18a43b39c046ae6d9aefb8296f5d146201d637.

Carreau commented on numpy/random/_generator.pyx:
```diff
@@ -5022,13 +5022,13 @@ def default_rng(seed=None):
     `Generator`. Here are several ways we can construct a random 
     number generator using `default_rng` and the `Generator` class. 
-    
+
```
I don't know if we want to have whitespace trimming in unrelated files.

agriyakhetarpal replied:
I'm happy to revert the change if needed. I just happened to notice it and thought it would be okay to fix. I'll move most of the changes to a separate PR in a moment as suggested above so that this PR can be limited to be related to just "enabling" the examples.

Carreau commented:
> About 90% of the diff here is adding `import numpy as np`

I have the impression that this would be an issue everywhere in the ecosystem. Do you belive an upstream global option that include a few lines at the beginning of all example (notebooks only, once clicked on "Try it in your browser!") would be useful ?

rgommers commented:
> I have the impression that this would be an issue everywhere in the ecosystem. Do you belive an upstream global option that include a few lines at the beginning of all example (notebooks only, once clicked on "Try it in your browser!") would be useful ?

I don't know that it is. We already changed it in SciPy to add the imports, and I'd expect most packages that are less closely related to NumPy to not have copied this "leave out one import" style. Quick check of usual suspects: scikit-learn has the imports everywhere, scikit-image has them in long-form examples but not docstrings, Pandas has a mix. I'd just let them make it fully consistent, rather than add more magic.

lesteve commented:
> I have the impression that this would be an issue everywhere in the ecosystem. Do you belive an upstream global option that include a few lines at the beginning of all example (notebooks only, once clicked on "Try it in your browser!") would be useful ?

Just to make 100%, by upstream you mean jupyterlite-sphinx, right? FWIW in sphinx-gallery we added the possibility to add content (both markdown and code) at the beginning of the JupyterLite notebook based on the notebook content, see the [doc](https://sphinx-gallery.github.io/stable/configuration.html#jupyterlite) for more details. Maybe you could imagine a similar mechanism in `jupyterlite-sphinx` where if `np.` is inside the code example you would add a `import numpy as np` to the JupyterLite notebook?

In the scikit-learn gallery examples, there are still quite a few reasons where we need to add Pyodide specific code. Off the top of my head (see https://github.com/scikit-learn/scikit-learn/blob/87fa6547d6a864c32851a811a7758837f47bb170/doc/conf.py#L579-L636 for all the details)
+ early-import some packages where needed e.g. pandas and matplotlib in scikit-learn. Only top-level imports work in Pyodide, for optional dependencies in scikit-learn like matplotlib and pandas we import them inside functions and the message error is a bit confusing.
+ you need patching with `pyodide_http` when downloading data
+ some packages need to be installed with `%pip`, i.e. are not packaged by Pyodide

melissawm commented:
From a user perspective, I think it's better to have the imports everywhere. This means every example can be copy-pasted and works without any extra cognitive load. For us it's repetition but most users will read one page at a time and may miss that information.

Carreau commented:
That is fair, so in the case we do actually want the explicit numpy import, is a long term plant to get rid of :

https://github.com/numpy/numpy/blob/26a2e6c882d25eefd447a25ba930dd3721505834/numpy/conftest.py#L123-L125 

This way doctests will fail if the example make use of `np` and it is not  explicitly imported This assumes of course that all doctests get turned into interactive examples, which may not be true for some implementation methods docs that do not end up un API docs.

agriyakhetarpal committed:
REL: Update release note to highlight in-browser support

[skip cirrus] [skip azp]

Co-authored-by: M Bussonnier <bussonniermatthias@gmail.com>

agriyakhetarpal committed:
DOC: Fix "Inexact types" example not rendering

[skip cirrus] [skip azp]

agriyakhetarpal committed:
DOC: Add margins to `try_examples_outer_iframe`

Co-Authored-By: M Bussonnier <bussonniermatthias@gmail.com>

agriyakhetarpal committed:
REL: Fix unneeded words and typos in release note

[skip cirrus] [skip azp]

agriyakhetarpal committed:
DOC: Fix indents in "N-dimensional enumeration" example

[skip cirrus] [skip azp]

agriyakhetarpal commented:
> However, this is technically kinda separate from supporting interactive docs, so I do have a preference for separating the two changes - because it makes it much easier to review and understand what we actually need to add for interactive docs.

@rgommers, done in gh-26814. Fortunately, my commit history meant that the changes were separate enough for me to not have issues with cherry-picking those commits to another branch. I shall rebase this branch once that PR is merged.

agriyakhetarpal commented:
> However, this is technically kinda separate from supporting interactive docs, so I do have a preference for separating the two changes - because it makes it much easier to review and understand what we actually need to add for interactive docs.
>
> Do you think you could try that @agriyakhetarpal?

@rgommers, done in gh-26814. Fortunately, my commit history meant that the changes were separate enough for me to not have issues with cherry-picking those commits to another branch. I shall rebase this branch once that PR is merged.

ngoldbaum commented:
Is there a way to somehow declare that an example is interspersed with text? That way we can avoid redefining arrays in the middle of examples like [where I'm commenting in the other PR](https://github.com/numpy/numpy/pull/26814/files#r1659183954).

agriyakhetarpal commented:
> Is there a way to somehow declare that an example is interspersed with text?

There doesn't seem to be a way so far, @ngoldbaum. However, if you add the related text for the example in a way that is indented inside the `..try_examples::` block at the same level as that for the related code, it sort of implies that it is, because the text gets included as a Markdown block interspersed between code blocks inside the JupyterLite-based notebook that the directive creates.

agriyakhetarpal commented:
To answer some other comments above from @melissawm:

> Do we want to have a functional jupyterlite kernel in CircleCI to verify that this is working on every PR? I'm not sure that's necessary. So I wouldn't spend too much time on trying to get that working. Happy to hear other thoughts though.

I don't think it's necessary either, because the extension more or less works in deployed documentation websites. However, I think @lesteve's solution of adding an `index.html` string in the URL is minimal and great, and I submitted a PR to fix this in `jupyterlite-sphinx`: https://github.com/jupyterlite/jupyterlite-sphinx/pull/182.

> I know we haven't necessarily figured out the details yet, but I feel like we should have some mention of the examples and jupyterlite in general in the contribution and maintenance docs. I would either link or get inspiration from [scipy.github.io/devdocs/dev/core-dev/index.html#wrapping-up](https://scipy.github.io/devdocs/dev/core-dev/index.html#wrapping-up) (for the relese guide) and the "Interactive documentation" section on [scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#rendering-documentation-locally-with-sphinx](https://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#rendering-documentation-locally-with-sphinx) . This can be done in a follow-up but in that case we should open an issue so we don't forget.

I agree with this proposition. If someone wishes to open a follow-up issue before I do, please do – thanks! Personally, I feel it can be included here itself once gh-26814 gets merged because the diff here will get smaller. I can add the corresponding notes in both these locations.

melissawm commented:
> Is there a way to somehow declare that an example is interspersed with text? That way we can avoid redefining arrays in the middle of examples like [where I'm commenting in the other PR](https://github.com/numpy/numpy/pull/26814/files#r1659183954).

You can do that with the `.. ipython::` directive. See, for example, here: https://raw.githubusercontent.com/numpy/numpy/9138d033e2d2a30163100a99fdeb02e97b9089be/doc/source/reference/routines.polynomials.rst and https://numpy.org/doc/stable/reference/routines.polynomials.html

I'm not sure how an `.. ipython::` directive would be rendered in the jupyterlite examples so maybe not the best approach anyway.

agriyakhetarpal committed:
Merge branch 'main' into interactive-docs

agriyakhetarpal committed:
DOC, DEP: Bump to `jupyterlite-sphinx` v0.16.2

[skip cirrus] [skip azp]

agriyakhetarpal commented:
I am quoting some of the previous conversations here to bring them back to the forefront after I have pushed my last two commits at the time of writing: 492473539ef523bd6908a31c3108cc5805f2ea48 and b31f5f2f0c4d7888298816b92932a1a95d3a40d7

> About 90% of the diff here is adding `import numpy as np`, to make the documentation examples more self-contained and copy-paste-able. It seems fine to me, and I'd prefer to just do this now that there's one more good reason for it, rather than bikeshed it again to the extent done for SciPy in https://github.com/scipy/scipy/issues/13049.

and

>> However, this is technically kinda separate from supporting interactive docs, so I do have a preference for separating the two changes - because it makes it much easier to review and understand what we actually need to add for interactive docs.

> @rgommers, done in [gh-26814](https://github.com/numpy/numpy/pull/26814). Fortunately, my commit history meant that the changes were separate enough for me to not have issues with cherry-picking those commits to another branch. I shall rebase this branch once that PR is merged.

<hr>

@rgommers, gh-26814 has now been merged, which addresses your previous comment about the demarcation between "making the examples work" and "enabling interactivity for examples" – the example code snippets are now self-contained in most locations. Therefore, this PR can now be revisited. I'll look at the changes again here since it has been a while, push some cleanup commits as necessary, and mark this ready for review.

agriyakhetarpal committed:
STY: Add newline at the end of `numpy.css`

[skip cirrus] [skip azp]

agriyakhetarpal committed:
STY: Fix punctuation for global warning text

agriyakhetarpal committed:
STY: Add EOF newline in TryExamples config file

agriyakhetarpal committed:
Merge branch 'main' into interactive-docs

agriyakhetarpal committed:
DOC: Remove some duplicate `import numpy` statements

[skip cirrus] [skip azp]

agriyakhetarpal committed:
DOC: Remove implicit assumption about import

[skip cirrus] [skip azp]

rgommers reviewed (commented):
Thanks Agriya! The diff looks quite clean to me now, and the rendered version looks polished and works as expected for me:

<img width="633" alt="image" src="https://github.com/user-attachments/assets/a380c26b-02da-40b5-bfef-e3e8b518de19">

* * *

To make sure: this is complete for the API reference guide? All examples that weren't already in regular doctest mode that is auto-detected now have a `.. try-examples::`, right?

The difference between 1.26 and 2.0 in behavior isn't _that_ large in the end, so I'm thinking that it's probably okay to merge this. Although having NumPy 2.0 land in-tree in Pyodide would be a lot more comfortable. Could you remind me of the expected timing for that?

agriyakhetarpal commented:
> To make sure: this is complete for the API reference guide? All examples that weren't already in regular doctest mode that is auto-detected now have a `.. try-examples::`, right?

Yes. The `.. try_examples::` directives that I have inserted manually here are needed just for those code snippets that are currently not under the "auto-detectable" heading as follows:

```rest
Examples
-------
```

For, say, `numpy/_core/code_generators/ufunc_docstrings.py` for example, the heading is present in all of the `add_newdoc()` function calls, so the examples in that file are rendered interactive without the need for the directive (through `global_enable_try_examples = True` in `conf.py`).

> The difference between 1.26 and 2.0 in behavior isn't _that_ large in the end, so I'm thinking that it's probably okay to merge this. Although having NumPy 2.0 land in-tree in Pyodide would be a lot more comfortable. Could you remind me of the expected timing for that?

I see – yes, I think it is okay to merge this, too (awareness about the unavailability of NumPy v2.0 features is, IMO, adequately raised in the global warning text for all of the examples). In the case of Pyodide, NumPy 2.0 in-tree (https://github.com/pyodide/pyodide/pull/4925) is complete, but it relies on SciPy v1.13.0 in-tree being complete (https://github.com/pyodide/pyodide/pull/4719). I have revisited that PR recently and I am working on it. Either way, @ryanking13 and I had agreed that NumPy v2 should not be available until v0.27.X because of the ABI break for NumPy, which would need version bumps for all existing packages (which I fixed – there were not many, thankfully). I don't think Pyodide 0.27.X is expected to land until the next quarter, i.e., Q4 2024, which is a considerable amount of time from now.

rgommers commented:
Thanks @agriyakhetarpal, that all sounds good. We're close to the 2.1.0 branch date, but this seems safe enough to merge. I'll give it another 24 hours or so in case anyone still wants to review or has concerns.

agriyakhetarpal committed:
DOC, REL: Update interactivity goal to NumPy v2.1.0

Co-Authored-By: Ralf Gommers <ralf.gommers@gmail.com>

agriyakhetarpal commented on doc/neps/roadmap.rst:
```diff
@@ -75,7 +75,8 @@ planned improvements. Adding more tutorials is underway in the
 
 We also intend to make all the example code in our documentation interactive -
-work is underway to do so via ``jupyterlite-sphinx`` and Pyodide.
+work is underway to do so via ``jupyterlite-sphinx`` and Pyodide. NumPy 2.0.0
```
Addressed in f1a15cc10f67f826632a4441427fa4b5cfc32ba0.

agriyakhetarpal commented:
Thanks for the review! In those 24 hours, I'll try to find edge cases programmatically based on the advice offered by @mattip in https://github.com/numpy/numpy/pull/26814#issuecomment-2248325688 and fix them as necessary. For instance, [the documentation preview for the "Random sampling" (`np.random`) page](https://output.circle-artifacts.com/output/job/9c92a67a-ce63-49ec-9525-bd9db8642f41/artifacts/0/doc/build/html/reference/random/index.html) in gh-26814 had improper indentation, which seems to have been fixed in [its preview for this PR](https://output.circle-artifacts.com/output/job/7af4aec3-951e-45b2-b220-8664a4169ebe/artifacts/0/doc/build/html/reference/random/index.html) supposedly due to `jupyterlite-sphinx`'s error correction abilities of sorts :), but it will still be good to double-check as much as possible.

agriyakhetarpal committed:
DOC: Add missing `np.strings.add` interactive example

agriyakhetarpal committed:
DOC: Disable a few non-relevant pages

Pages such as those related to packaging
via `numpy.distutils`, `typing`, and `f2py`
are not necessarily required to run in
interactive mode and do not offer much in
terms of examples to the user in general.

agriyakhetarpal committed:
DOC: Remove code block directives (`_typing/_add_docstring.py`)

[skip azp] [skip cirrus]

agriyakhetarpal committed:
DOC: Fix docs build failure from rubrics warnings

This reverts commit 905279a38266c43a7a0ed78c163c2690860184f5.

[skip azp] [skip cirrus]

agriyakhetarpal committed:
DOC: Fix indentation for `np.strings.add()` example

agriyakhetarpal committed:
DOC: Clean up file paths to be ignored

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal commented:
I'm marking this PR as a draft for now and writing this makeshift message so that I can edit it later in the day to add more information – I have a couple of questions and comments that I (and the rest of the reviewers here) can address or discuss.

agriyakhetarpal commented:
As noted, I have a list of comments and questions that I (and the rest of the reviewers here) can address or discuss:

1. I went through all of the sections in the table of contents, ranging from "Data type routines" to "Linear algebra" to "Mathematical functions with automatic domain" and more. I have found that the buttons for almost all of the examples render fine in general, but I haven't tested or found a way to display each example – that's laborious and most likely outside the scope of this PR. Maybe `pytest-doctestplus` can be supported in Pyodide someday for this.
2. I disabled the examples on the packaging (`numpy.distutils`) and `numpy.typing` pages through the `try_examples.json` configuration file, since interactive examples for those sections are not really relevant, of course. The examples under "Input and output" are expected to fail in many places because of the lack of a file system in the browser and with intricacies in saving and loading files, and so will those for many of the examples under `np.strings`. Should I disable both of them as well for now, and the latter, at least, can be enabled when NumPy v2 is available in Pyodide and the docs can be rebuilt for that purpose?
3. This page: https://output.circle-artifacts.com/output/job/1d6ca46d-6e58-40cb-bb1c-b9387dddfa6f/artifacts/0/doc/build/html/reference/generated/numpy.finfo.html displays a long list of attributes after the Examples section. Would it be better to move the Attributes section before the Examples section, maybe just for the ploy that anything interactive should be preceded by the rest of the information necessary to enable the example?
4. Not that it is relevant here much because I disabled the examples on this page anyway, as noted in point 2, but https://output.circle-artifacts.com/output/job/1d6ca46d-6e58-40cb-bb1c-b9387dddfa6f/artifacts/0/doc/build/html/reference/typing.html#api has a malformed example with improper indentation which caught my regex's eye, and I haven't been able to figure out the reason. The docstring of the example itself is correct, and so are the ones for the examples that come after it (I confirmed by reordering them), so I'm not sure where the extra blockquote appears from. I can take a look at it after this PR if needed.
5. I'm not entirely happy with the contrast (or lack of it, rather) between the example buttons and the text within them when browsing in dark mode. It seems to be compromising readability:

	<details>
	<summary>Tap to show button in dark mode</summary>

	<img width="676" alt="Documentation for numpy.finfo in dark mode, displaying the interactive examples button with the text 'Try it in your browser'" src="https://github.com/user-attachments/assets/5add7e45-fcc1-4be1-afd8-994d8a6cc483">


	</details>

	I ran it through the WCAG compliance checker available at https://webaim.org/resources/contrastchecker/ and this is the 	result:

	<details>
	<summary>Tap to show results</summary>

	<img width="658" alt="Results from an accessibility audit of the interactive examples button under dark mode, that showcases how poor the results are" src="https://github.com/user-attachments/assets/647f74e1-e607-4e86-938f-6acfe543ce2a">

	</details>

	I think I'll need to use some of the other offerings by the PST instead of `var(--pst-color-info)`. Let me change this in a moment.

6. While checking the buttons on some of the pages, I noted that some of them, i.e., those for `numpy.strings.isalpha`, `numpy.strings.rfind`, `numpy.strings.startswith`, and `numpy.strings.mod` – don't contain any examples in their docstrings, and I don't see an open issue about this either. Would it be okay to add them? I could do all four of them in a follow-up PR.
7. Lastly, as noted in the PR description, I haven't enabled interactivity in the pages in the "User guide" section because of the nuance involved with its long-form-style content and the extra work required to make that work, but this page: https://output.circle-artifacts.com/output/job/1d6ca46d-6e58-40cb-bb1c-b9387dddfa6f/artifacts/0/doc/build/html/user/basics.rec.html has a few interactive examples that were automated – we should skip this page as well in `try_examples.json` for now, maybe?

rgommers commented:
Thanks Agriya for the attention to details. My thoughts:

1. makes sense to leave that out of scope of this PR
2. yes, disabling `np.strings` for now sounds good 
3. It won't help to move the Examples and Attributes sections around. The order is determined by `numpydoc` and the same for all classes (e.g., see `numpy.ndarray`). just leave it as it is now I'd say.
4. I don't see the problem so quickly from looking at that page. All looks okay to me.
5. Good catch. Using a standard PST color here makes sense. And ideally a follow-up will upstream as much as possible to PST so there's no need for any custom CSS in `numpy` or another project for the PST + jupyterlite-sphinx combo.
6. Yes. Not really a problem, that will happen for more docstrings. But feel free to open a separate PR for it.
7. I'd leave it alone. It seems to pull in docstrings, so doing nothing is preferred here I think (less work, smaller diff).

melissawm commented:
For 3, note this PR which I believe will fix this: https://github.com/numpy/numpydoc/pull/571

agriyakhetarpal committed:
DOC: Fix dark text for interactive example buttons

[skip azp] [skip cirrus] [skip actions]

agriyakhetarpal committed:
DOC: Make `numpy.strings` pages non-interactive

and properly escape regex patterns for other pages.

[skip azp] [skip cirrus] [skip actions]

agriyakhetarpal commented:
Thanks for the feedback! I fixed the accessibility issue in 6565c9df6aefc704e3ec78cbe2699ed4a39d58e0 (see the [passing results from the WCAG checker](https://webaim.org/resources/contrastchecker/?fcolor=000000&bcolor=79A3F2)) and addressed point 2 in 36b5e89ddd9670955a68cddec12be5c2ad3e9e81, so this is good to go now.

Regarding point 5: @steppi, would it be nice to add a default style for the buttons downstream to `jupyterlite-sphinx` that can claim compatibility with PST, i.e., to update/deprecate `:example_class:` and include a default set of colours to ship with our code that others can inherit for their buttons? We can add `--jupyterlite-sphinx-` as a prefix to our styles so that our variables don't conflict with those from other extensions, akin to `sphinx-design`, which does this with `--sd-` (and PST does so too, with `--pst-`). Our default style is just a white background and black text in a rectangle with no rounded corners, currently. I don't think it won't be too much of a maintenance burden with each advancing PST version. If you agree, I can put together an issue/PR to `jupyterlite-sphinx`, and we can work out the details there?

About point 4, this is the issue in particular, @rgommers:

<details>
<summary>Tap to view (this is a large image, please open it in a new tab and zoom it if it's not clear enough)</summary>

![output circle-artifacts com_output_job_1d6ca46d-6e58-40cb-bb1c-b9387dddfa6f_artifacts_0_doc_build_html_reference_typing html(iPad Air)](https://github.com/user-attachments/assets/3b132315-a825-4aa5-b038-db354f9e0d07)

So, the first example snippet under the "API" heading has an extra blockquote, but the one immediately after it does not.

</details>

<hr>

A few things for me to do after this PR gets merged will be to:

- address point 6 above – add some examples that were missing; and
- open an issue with the "DX" label (or should I open a PR directly?) to add instructions and notes about interactive examples for developers, as suggested by @melissawm in an earlier review comment: https://github.com/numpy/numpy/pull/26745#pullrequestreview-2133346535, and
- look into point 1, which has now been tracked in a discussion outside of this PR today and remains in the mix

agriyakhetarpal committed:
DOC: Better selector ordering for dark mode

steppi commented:
> Regarding point 5: @steppi, would it be nice to add a default style for the buttons downstream to `jupyterlite-sphinx` that can claim compatibility with PST, i.e., to update/deprecate `:example_class:` and include a default set of colours to ship with our code that others can inherit for their buttons?

I think having a better default sounds like a good idea but I don't really know enough about front end development to know what the best thing to do is here. The current means of styling the button with custom css and the existence of `:example_class:` were @Carreau's ideas, so he might have suggestions.

agriyakhetarpal committed:
Merge branch 'main' into interactive-docs

agriyakhetarpal committed:
Bump to `jupyterlite-sphinx` version 0.16.4

[skip azp] [skip cirrus] [skip actions]

agriyakhetarpal commented:
@Carreau, I think the earlier changes @steppi and I discussed in jupyterlite/jupyterlite-sphinx#199 were the correct ones, and we would need to cut a v0.16.5 release, based on the logs here: https://app.circleci.com/pipelines/github/numpy/numpy/28169/workflows/3d81381c-517e-4012-9a51-eaa1e66e0b28/jobs/41889 – it's returning a `str` object instead of a `_PathStr` one.

agriyakhetarpal committed:
Bump to `jupyterlite-sphinx` v0.16.5

Carreau commented:
> I think having a better default sounds like a good idea but I don't really know enough about front end development to know what the best thing to do is here. The current means of styling the button with custom css and the existence of `:example_class:` were @Carreau's ideas, so he might have suggestions.

Example class was just to avoid stying the button directly. And note that it can be set on a per-button, or globally. I think that right now the default is None, I have no objection making the default not None, but I'm not sure it is useful. I think the class might not even be always needed as the buttons to have `.try_examples_button_container` and `.try_examples_button` class already set. So it's more an override than anything else.

I'm worries of providing too many default and reusable values in something which main goal is not appearance but functionality. Usually it's better to provide the right hooks and let theme deal with that. Like PST to provide a `pst-example-button` and instruct to set `try_examples_global_theme='pst-example-button'`, or pst could already overrwrite `.try_examples_button_container` and `.try_example_button`.

So I don't think there is much we need to change.

agriyakhetarpal committed:
Bump to `jupyterlite-sphinx` >=0.17.1, `jupyterlite-pyodide-kernel` 0.4.7

agriyakhetarpal committed:
Update milestone for interactive documentation

agriyakhetarpal committed:
Merge branch 'main' into interactive-docs

agriyakhetarpal committed:
Use Pyodide kernel from `pip` in the meantime

agriyakhetarpal committed:
NumPy strings API examples should now work

agriyakhetarpal commented:
Hi all, with the release of [Pyodide version 0.27.0](https://github.com/pyodide/pyodide/releases/tag/0.27.0) comes NumPy 2.0.2 (via https://github.com/pyodide/pyodide/pull/4925). With https://github.com/jupyterlite/pyodide-kernel/pull/154 and [`jupyterlite-pyodide-kernel` version 0.4.7](https://github.com/jupyterlite/pyodide-kernel/releases/tag/v0.4.7), we now have Pyodide 0.27 working in JupyterLite. We can update the Pyodide kernel to 0.5.0 when we have a stable release (there is a 0.5.0b1 available at the time of writing). There is also a chance of updating to NumPy version `2.2.1` at some point in time, perhaps with a Pyodide 0.27.1 release – we are considering the possibility of having more frequent Pyodide 0.27.X releases to bring in more package updates from time to time, and subsequently, unvendoring all packages from the Pyodide runtime with 0.28 later in the year so that they can exist independently instead of relying on a Pyodide release.

I've made all these additions in the most recent commits above, and the deployment seems to be working, with no incumbent issues from a fresh look. I've re-enabled the NumPy 2.0 strings API examples, since they should work, too.

This is ready for review again, and the next steps for us should be to address gh-27310, and get a prototype of two pages under the NumPy user guide, i.e., "NumPy quickstart" and "NumPy: the absolute basics for beginners", rewritten and rendered as MyST in the form of long-form notebooks, in an attempt similar to scipy/scipy#20303 post PyWavelets/pywt#741 is finalised.

Another interesting, medium-term avenue is to explore [MyST-JS's JupyterLite integration](https://mystmd.org/guide/integrating-jupyter#jupyterlite) with [the `numpy-tutorials` website](https://numpy.org/numpy-tutorials)'s [MyST-JS deployment](https://github.com/numpy/numpy-tutorials/pull/212) [branch](https://github.com/numpy/numpy-tutorials/tree/mystjs), while simultaneously working on outstanding issues there to get it to be the default one.

agriyakhetarpal committed:
Disable JupyterLite source maps for smaller builds

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal commented:
d818f8c2f77e59d59c56d46ce7ee31bdd5ef393b disables [source maps for JupyterLite](https://jupyterlite.readthedocs.io/en/stable/reference/api/py/jupyterlite-core.html#jupyterlite_core.config.LiteBuildConfig.no_sourcemaps), based on scikit-learn/scikit-learn#26246. This brings the size of the JupyterLite build in the `doc/build/lite/` folder from 63.3 MiB (out of the total docs size of 180.7 MiB) to 22.4 MiB (with a total docs size of 140.2 MiB), which is a 22.4% reduction in build size. While doing this has the disadvantage of making debugging _slightly_ harder for end users, it is favourable for the aspect of saving space, especially for GitHub Pages and https://github.com/numpy/doc. I've employed the same configuration for sympy/sympy#27419 as well.

agriyakhetarpal committed:
`jupyterlite-pyodide-kernel` 0.4.7 is now available on conda-forge

This reverts commit ef3f14b13ab539078dab72b5b38d282d0780d8aa.

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal committed:
Bump `jupyterlite-sphinx` and associated Pyodide kernel

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal committed:
Bump JupyterLite dependencies in docs reqs as well

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal commented:
Hi @rgommers, we published https://github.com/jupyterlite/jupyterlite-sphinx/releases/tag/v0.18.0 some moments ago (thanks @steppi!), and I've updated the versions of `jupyterlite-sphinx` and `jupyterlite-pyodide-kernel` in 07139948abe35e1e26c79d6cc384aedd7736107b and 369e27d4463651849bc2e386c46253dbe72d37aa – we are good to proceed now, thanks!

agriyakhetarpal committed:
Fix incorrect constraint for `jupyterlite-sphinx`

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal committed:
Bump to Pyodide kernel 0.5.1; brings Pyodide 0.27.1

[skip actions] [skip azp] [skip cirrus]

agriyakhetarpal commented:
Commit e7972a54506e5f1a73f3b363892a3a9fdd8ae698 bumps to https://github.com/jupyterlite/pyodide-kernel/releases/tag/v0.5.1 that was released today, which brings the recent https://github.com/pyodide/pyodide/releases/tag/0.27.1.

(Apologies for the force push!)

melissawm reviewed (commented):
Hi @agriyakhetarpal ! The changes look good to me, I've successfully built this PR locally and it seems to work (although the jupyterlite blocks don't work locally).

The CircleCI version of the "Try it" works fine, but when I click "Open in new tab" I see this:

![Captura de imagem_20250121_141043](https://github.com/user-attachments/assets/c6e79b5d-c4d7-4f91-a5fa-f57885a9a05c)

Is this expected? (Maybe it's because it's in CI)

agriyakhetarpal commented:
Hi @melissawm, thanks for the review! I don't think I can reproduce it with CircleCI. Could you please retry with incognito mode, or with a different internet connection or mobile data if you can? This is what I get:

<img width="1512" alt="JupyterLite notebook interface that displays an example from the NumPy documentation, more specifically, np.ma.HardMask, running in a kernel provided by the Pyodide project" src="https://github.com/user-attachments/assets/b9375222-444d-4b96-b941-57c98b89c0b5" />

<br>
<br>

For reference, I tried this link: https://output.circle-artifacts.com/output/job/4b11fff4-e43b-4010-af4e-de2e84deed66/artifacts/0/doc/build/html/lite/notebooks/index.html?path=88efbc39_8e7b_4e21_86c0_519bb66f4f6a.ipynb

> although the jupyterlite blocks don't work locally

They should work locally as well. Could you please share any errors or output that you might be seeing?

agriyakhetarpal committed:
Merge branch 'main' into interactive-docs

agriyakhetarpal committed:
Bump to `jupyterlite-pyodide-kernel` 0.5.2

[skip actions] [skip azp] [skip cirrus]

melissawm reviewed (approved):
Now it works fine! Thanks @agriyakhetarpal

rgommers reviewed (approved):
This looks pretty clean, nice work @agriyakhetarpal! Time to give this a go. Thanks to @melissawm and everyone else who contributed as well.

@agriyakhetarpal could you update us on the next steps here? I think it includes upgrading to NumPy 2.2 in Pyodide and enabling use of nightlies, and probably also enabling functionality on the User Guide? Anything else?

agriyakhetarpal commented:
> @agriyakhetarpal could you update us on the next steps here? I think it includes upgrading to NumPy 2.2 in Pyodide and enabling use of nightlies, and probably also enabling functionality on the User Guide? Anything else?

Thanks, @rgommers! The following steps here are indeed those, and also, we need to add some docs to resolve gh-27310, similar to @melissawm's SciPy PR (https://github.com/scipy/scipy/pull/22385, which I'll be reviewing and we can adopt similar messaging for NumPy when done).

- Upgrading to NumPy 2.2 in Pyodide was last blocked due to https://github.com/numpy/numpy/issues/27865, but I can revisit it.
- I had a branch for starting similar work the User Guide, but it's outdated now due to our relatively more recent changes in our approach(es). I can get started on converting the pages to MyST, though I would propose that we wait for https://github.com/ipython/ipython/issues/14591 before we merge it.
    - It probably makes sense for me to port one or two pages in a PR before getting to the rest of them all – I'd like to start with the ["NumPy quickstart"](https://numpy.org/doc/stable/user/quickstart.html) page before proceeding to ["NumPy: the absolute basics for beginners"](https://numpy.org/doc/stable/user/absolute_beginners.html), as the latter is much longer.

rgommers commented:
Thanks Agriya. Docs would be a nice next step here indeed, leveraging the work Melissa did for SciPy (the setups and considerations are almost identical). It'd also be fine with me to have a shorter explanation and link to the SciPy docs on the topic for more details.

> Upgrading to NumPy 2.2 in Pyodide was last blocked due to [...]

This would be good to prioritize. It seems like there was a way forward (or two possibly) there, by reverting making symbols private for example.

Re User Guide: agreed with waiting for better tracebacks and doing 1-2 pages first.
